zrml/prediction-markets/src/migrations.rs

```diff
@@ -15,6 +15,495 @@
+const PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION: u16 = 5;
+const PREDICTION_MARKETS_NEXT_STORAGE_VERSION: u16 = 6;
+
+const AUTHORIZED: &[u8] = b"Authorized";
+const AUTHORIZED_OUTCOME_REPORTS: &[u8] = b"AuthorizedOutcomeReports";
+
+// TODO increase storage versions here, when last migration was executed
+const AUTHORIZED_REQUIRED_STORAGE_VERSION: u16 = 1;
+const AUTHORIZED_NEXT_STORAGE_VERSION: u16 = 2;
+
+pub struct UpdateMarketIdsPerDisputeBlock<T>(PhantomData<T>);
+
+// Delete the auto resolution of authorized and court from `MarketIdsPerDisputeBlock`
+impl<T: Config + zrml_market_commons::Config + zrml_authorized::Config> OnRuntimeUpgrade
+    for UpdateMarketIdsPerDisputeBlock<T>
+{
+    fn on_runtime_upgrade() -> Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let prediction_markets_version = StorageVersion::get::<Pallet<T>>();
+        if prediction_markets_version != PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION {
+            log::info!(
+                "prediction-markets version is {:?}, require {:?}",
+                prediction_markets_version,
+                PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION,
+            );
+            return total_weight;
+        }
+        log::info!("UpdateMarketIdsPerDisputeBlock: Starting...");
+
+        let mut new_storage_map = Vec::new();
+        let mut authorized_ids = Vec::new();
+        for (key, mut bounded_vec) in storage_iter::<BoundedVec<MarketIdOf<T>, CacheSize>>(
+            PREDICTION_MARKETS,
+            MARKET_IDS_PER_DISPUTE_BLOCK,
+        ) {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().reads(1));
+
+            bounded_vec.retain(|id| {
+                if let Ok(market) = <T as crate::Config>::MarketCommons::market(id) {
+                    match market.dispute_mechanism {
+                        MarketDisputeMechanism::Authorized => {
+                            authorized_ids.push(*id);
+                            false
+                        }
+                        MarketDisputeMechanism::Court => false,
+                        MarketDisputeMechanism::SimpleDisputes => true,
+                    }
+                } else {
+                    // no market for id in MarketIdsPerDisputeBlock
+                    false
+                }
+            });
+
+            new_storage_map.push((key, bounded_vec));
+        }
+
+        for (key, new_bounded_vec) in new_storage_map {
+            put_storage_value::<BoundedVec<MarketIdOf<T>, CacheSize>>(
+                PREDICTION_MARKETS,
+                MARKET_IDS_PER_DISPUTE_BLOCK,
+                &key,
+                new_bounded_vec,
+            );
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        }
+
+        let now = <frame_system::Pallet<T>>::block_number();
+        for id in authorized_ids {
```

Comment:
With the above call to `retain` I delete all `Authorized` ids out of `MarketIdsPerDisputeBlock`.

After that the idea was to resolve every `authorized` market after the runtime upgrade until `ReportPeriod` is over. Otherwise `authorized` markets would never resolve after this update. When there is no report by the authority, it resolves to the oracle outcome. Because the other storage migration below is used to save an additional `resolve_at`, I used the same approach there.

If the authority reports, the `resolve_at` storage gets overwritten and the authority has `CorrectionPeriod` time until it resolves.

> The authority has not yet reported. In this case, we should not queue the resolution.

For this case I like the idea to allow `ReportPeriod` until it resolves. When we would not queue the resolution, we end up with the situation, that we could wait forever. We should not rely on the committee. One should use the fail-safe approach. As discussed yesterday in the call.

> The authority has reported. In this case, we should resolve the market either as soon as possible, or maybe after the CorrectionPeriod.

This case is not regarded by me. Good point. I expected, that the authority did not report in every case yet. I could check the storage `AuthorizedOutcomeReports` and if it is present, then I should use `CorrectionPeriod` as you suggested. Otherwise the authority needs to wait `ReportPeriod` and could only shorten this with an active call to `authorize_market_outcome` again.

Reply:
Turns out, that this not easy at all. Because I have another storage migration and need to query the storage of `authorized` inside `zrml_prediction_markets` migration. I would be really in favour of the current solution, otherwise it gets even more complex. I mean the authority can act, when the period is too long for them, but don't have to, because it resolves automatically.

Reply:
> For this case I like the idea to allow `ReportPeriod` until it resolves. When we would not queue the resolution, we end up with the situation, that we could wait forever.

Why? Doesn't `has_failed` return `true` at some point?

> Turns out, that this not easy at all.

I understand. What's important is that the two migrations are consistent with one another. Less important, but still something to keep an eye on is that the "old" reports are migrated in a fashion that is close to what would have happend to them if they had been submitted after these changes.

Why does the `authorized` migration set `resolve_by` to `now` plus `ReportPeriod` when it could set it to `now` plus `CorrectionPeriod`?

Reply:
> Why? Doesn't has_failed return true at some point? 

We decided to make the authority infallible. Let us stay with that decision for now.

> Less important, but still something to keep an eye on is that the "old" reports are migrated in a fashion that is close to what would have happend to them if they had been submitted after these changes.

Yeah, so the best would be to have the same behaviour as before for authorized markets. Let me think about that, how this would impact the storage migration and behaviour...

> Why does the authorized migration set resolve_by to now plus ReportPeriod when it could set it to now plus CorrectionPeriod?

Because `ReportPeriod` is much longer, so that the authority has enough time before the resolution happens.

Reply:
On battery station I get this error for the migration: `For market id "[202, 165, 87, 63, 249, 172, 218, 231, 176, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]" storage iter should only find present (Option::Some) values`

Do you know what this could be?

Test was made after this commit 8b81a11ac88978d19f2276942a21ae23e8ad0c8c.

Reply:
Could it be, that the `on_resolution` was called, but the market id is still inside the storage for `AuthorizedOutcomeReports`? Because we have an old storage value from the past.

So, then I would remove the panic call in try-runtime.

Reply:
Solved in private communication. Turns out that market has broken state due to a bug fixed in #701.